### PR TITLE
[ROCm] Add MIOpen API-based tuning for Batchnorm

### PR DESCRIPTION
**Pull request by dnikolaev-amd**
## Summary

Enable MIOpen API-based kernel tuning for batch normalization on ROCm when `torch.backends.cudnn.benchmark` is enabled 

- Add `MiopenTuningPolicyGuard` wrapper in `BatchNorm_miopen.cpp` that sets `miopenTuningPolicySearch` on the MIOpen handle before forward/backward batch norm calls and restores the previous policy afterward.
- Add compile-time guard `MIOPEN_HAS_TUNING_POLICY` in `miopen-wrapper.h` so builds against MIOpen headers older than 3.6 do not reference `miopenGetTuningPolicy` / `miopenSetTuningPolicy`.
- Add ROCm-only test `test_batchnorm_miopen_tuning` that verifies tuning policy is applied via MIOpen logs.

## Motivation

MIOpen introduced handle-level tuning policy APIs (`miopenGetTuningPolicy` / `miopenSetTuningPolicy`) to control kernel search/tuning. Convolution already benefits from benchmark-driven tuning via `miopenSetConvolutionFindMode`; batch norm had no equivalent hook. This change wires batch norm into the same `torch.backends.cudnn.benchmark` flag used on ROCm.

## Implementation

**`aten/src/ATen/miopen/miopen-wrapper.h`**
- Define `MIOPEN_HAS_TUNING_POLICY` when `MIOPEN_VERSION >= 3.6`.

**`aten/src/ATen/native/miopen/BatchNorm_miopen.cpp`**
- `MiopenTuningPolicyGuard` sets tuning policy based on `at::globalContext().benchmarkCuDNN()`:
  - `benchmark=True` -> `miopenTuningPolicySearch`
  - `benchmark=False` -> `miopenTuningPolicyNone`
- Calls `miopenGetTuningPolicy` on every forward/backward; calls `miopenSetTuningPolicy` only when the desired policy differs from the current one.
- Guard is installed in both `miopen_batch_norm` (forward) and `miopen_batch_norm_backward`.
- Version gating by `#if MIOPEN_HAS_TUNING_POLICY`

## Why `miopenGetTuningPolicy` is called on every batch norm

The MIOpen handle is long-lived and shared across ops, but `torch.backends.cudnn.benchmark` can change at any time via nested Python contexts. We read the current handle policy on each batch norm call so we can:

1. **React to the current `benchmarkCuDNN()` value** — evaluated at call time, not cached.
2. **Skip `miopenSetTuningPolicy` when already correct** — compare desired vs current policy before setting.
3. **Restore the previous handle policy in the destructor** — save what was on the handle before we potentially changed it.

Nested contexts illustrate why per-call reads are necessary:

```python
with torch.backends.cudnn.flags(enabled=True, benchmark=True):
    # benchmark=True -> MIOpen tuning ON
    bn(x)

    with torch.backends.cudnn.flags(enabled=True, benchmark=False):
        # benchmark=False -> MIOpen tuning OFF
        bn(x)

    # benchmark=True again (outer context restored)
    bn(x)

# original benchmark value restored
bn(x)
```

Each `bn(x)` may need a different tuning policy depending on the active context. Because the handle outlives any single op, we cannot assume its policy still matches `benchmarkCuDNN()` from a prior call.

## Test plan

- [x] `lintrunner -a aten/src/ATen/miopen/miopen-wrapper.h aten/src/ATen/native/miopen/BatchNorm_miopen.cpp test/test_nn.py`
- [ ] ROCm: `python test_nn.py -v -k test_batchnorm_miopen_tuning`
- [ ] ROCm: batch norm test suite passes (e.g. `python test_nn.py -v -k test_batchnorm`)

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Review by jeffdaily — CHANGES_REQUESTED** [review]
Thanks @dnikolaev-amd, the overall approach looks right and matches the conv find-mode precedent. A few issues to address:

**Destructor can call `std::terminate` (must fix).** `~MiopenTuningPolicyGuard` uses `MIOPEN_CHECK`, which throws (`aten/src/ATen/miopen/Exceptions.h`). Destructors are implicitly `noexcept`, so any failure in the restore call terminates the process. It also runs during stack unwinding when the guarded `miopenBatchNormalization*` call itself throws, which is exactly when the handle is most likely to be unhappy. Please check the status manually and `TORCH_WARN` on failure instead of throwing.

**Guard placement.** The struct is defined just after the anonymous namespace closes, giving it external linkage at `at::native` scope. Move it into the anonymous namespace next to `expandScale`, and delete the copy/move constructors so an accidental copy can't double-restore.

**Test gates on the wrong version.** The skip uses `torch._C._cudnn.getRuntimeVersion()`, but `MIOPEN_HAS_TUNING_POLICY` is decided at compile time from the build headers. A torch built against 3.5 headers running on a 3.6 runtime passes the skip and then fails `assertIn("miopenSetTuningPolicy", ...)`. Gate on `torch._C._cudnn.getCompileVersion()` (already bound in the same shim).

**The `benchmark=False` leg can pass vacuously.** If `MIOPEN_ENABLE_LOGGING` stops emitting, logging moves off stderr, or BN stops dispatching to MIOpen entirely, `assertNotIn` passes without testing anything. Add a positive control to both runs, e.g. `assertIn("miopenBatchNormalizationForwardTraining", result.stderr)`, plus `assertIn("miopenGetTuningPolicy", ...)` on the `False` leg to prove the guard executed.

**Search vs SearchDbUpdate.** The MIOpen header contradicts itself on `miopenTuningPolicySearch`: the enum-group doc says "tune and update the database" while the inline value doc says "tune but do not update the db". If Search doesn't persist to PerfDb, every process re-pays the tuning cost and `miopenTuningPolicySearchDbUpdate` is probably what benchmark mode wants. Can you confirm the actual semantics and document the choice in a comment?

**Restore defeats the skip-set optimization.** Because the destructor always restores, in the `benchmark=True` steady state every call does Get + Set(Search) + Set(None); the "skip when already correct" path only ever triggers with `benchmark=False`. Cheap host calls so this is fine, but the PR description overstates it. If the restore is deliberate (don't leak Search to other ops on the shared handle), a short comment saying so would help.

Two questions:
- Is 3.6 the right floor? I've seen a ROCm nightly SDK ship a header stamped 3.5 that already contains `miopenGet/SetTuningPolicy`. If the API ships in some 3.5.x release, those builds silently compile the feature out.
- The ROCm items in the test plan are unchecked and CI images likely carry MIOpen < 3.6, so the new test will skip everywhere in CI. Can you paste a passing run from a MIOpen 3.6 machine?

*This review was drafted with the assistance of an AI tool; I have reviewed and edited it.*
